compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -174,6 +174,14 @@ object Implicits:
                 migrateTo3
                 && tpw.derivesFrom(defn.FunctionClass(1))
                 && ref.symbol != defn.Predef_conforms
+
+              if isFunctionInS2 then
+                val span = ref.symbol.span
+                report.migrationWarning(
+                  s"Functions will no longer work as implicit conversions in Scala 3. Use an implicit parameter of Conversion type instead.",
+                  ctx.source.atSpan(span)
+                )(using irefCtx)
+
```

Comment:
A potential issue: how can we know that the symbol is defined in the current source file?

Could you create a test that involves two separately compiled files? Here is an example on how to organize the tests: https://github.com/lampepfl/dotty/tree/master/tests/neg/inlineAccess

Reply:
Actually we can't and I realised that when one of the community build tests failed. I simply found this line working in some other place and used it because I was not aware that the source position could be taken directly from a symbol.

Now I'm trying to figure out how the message should actually be displayed so that it's understandable for a user what should be changed in the code to fix the issue. If the conversion is defined within the user's code that should be quite straightforward - we warn on the line with the definition (which is either in the file where the conversion is applied or another). But if the conversion was defined in an external library we might get no source position at all. And even if we got the position from an external library, one normally has no possibility to modify it so to remove the warning message one would have to apply the conversion explicitly in their code (which might be quite problematic when e.g. the code heavily relies on some DSL) and in such a case it would make sense to show the warning at the point of conversion application. But this would introduce a not very nice asymmetry with the case when we get the source position of the definition.

Currently I'm investigating into the case when conversions are derived by chaining implicits because it looks like the source position might not be available then even if everything is defined within the same codebase.

Reply:
According to https://github.com/lampepfl/dotty/pull/9935 , the policy has changed:

> Do not warn if an old-style conversion is used

So, maybe it's fine to not issue a warning here?

Reply:
After reading that PR I'm a bit confused what is understood as "old-style conversion".
There seem to be 3 ways to implicitly convert a value of type A to a value of type B:
```
/* 1 */ implicit def a2b(a: A): B = ???
/* 2 */ implicit val a2b: A => B = ???
/* 3 */ implicit val a2b: Conversion[A, B] = ???
```
According to the latest documentation https://dotty.epfl.ch/docs/reference/changed-features/implicit-conversions-spec.html in Scala 3.0 the case 2 will not work anymore unless compiled with `-source:3.0-migration` flag. #9935 refers to the case 1 as `old-style implicit def conversion`. Does this mean it has been already decided that the case 1 will be deprecated soon? I've seen the discussion here https://contributors.scala-lang.org/t/can-we-wean-scala-off-implicit-conversions/4388 but it doesn't seems to reach any final conclusions. Anyway my changes are related specifically to `-source:3.0-migration` flag. #9935 limits an overwhelming number of warnings on a feature which is potentially dangerous but in most cases will work as expected. My changes are supposed to warn on things that would normally stop compiling in Scala 3 to smoothen the migration process. So I assume users using the migration flag would like to get these warnings.

To make the warnings consistent no matter which of the cases mentioned in the previous comment occurs, I might try to consistently use the source position of implicit conversion application and include the fully qualified name of the val/def working as the conversion. Then the user will have a choice to either update the definition of the conversion or make the conversion explicit.

Reply:
Thanks for the detailed analysis, @prolativ , it makes perfect sense 👍 

BTW, the method `Symbol.isDefinedInCurrentRun` might be useful.

Reply:
Warnings are now consistently reported in the place of implicit conversions applications.
Proper test cases were added.